### DISCUSSION TITLE
@accounts/password username validation

### DISCUSSION THREAD

**Discussion by MohanNarayana** [accounts-js/accounts#1213] [Q&A]
The username validation is done against this regex pattern `/^[a-zA-Z][a-zA-Z0-9]*$/`
Is there a reason why username must start with a letter? I feel it should accept any alphanumeric starting with a letter or number(all numbers too). Is changing this validation a consideration?

Thanks,
Mohan

**Comment by pradel** [comment]
Changing the regex is not really an option as it would be a breaking change for existing users. However you can use your own regex easily with this option https://www.accountsjs.com/docs/api/password/interfaces/accountspasswordoptions#optional-validateusername

**Reply by MohanNarayana** [reply]
Thanks for the work around. I am still curious to know the reasoning behind going with the above pattern for username.

**Reply by pradel** [reply]
I think at the time I just took the same regex that meteor had. As it was a long time ago I am not 100% sure where this is coming from :)

**Reply by MohanNarayana** [reply]
I went with the above suggestion for custom username validation but I am not getting it to work all the way. Basically I want the username to be any alphanumeric string.

```
const password = new AccountsPassword({
    validateUsername: async (username) => {
      const usernameRegex = /^[a-zA-Z0-9]*$/;
        if(!(typeof username === 'string' && usernameRegex.test(username))){
          return false;
        }
      return true;
    }
  });
```

I am able to override the default validation but it does not stop user creation for an invalid user. It goes into the `if` condition and ends up creating an invalid user. 

Firstly, is my code correct?
How can I stop the user creation and throw an error instead? (I tried throwing an error but the user creation went through and I got a `Unhandled promise rejection`)

I appreciate any help.
Thanks

**Reply by pradel** [reply]
Okay that's strange, can you try to use the `validateNewUser` option instead? You can do the validation for the username like in this example https://www.accountsjs.com/docs/strategies/password#extend-the-user-and-validate-custom-fields

**Reply by MohanNarayana** [reply]
I tried `validateNewUser`. It does not override the default validation. Maybe it only validates the extended fields?
With `validateUsername` I could see the validation override. Should `validateUsername` return an error or boolean? Or should it return nothing?

**Reply by pradel** [reply]
This is the code responsible for the check https://github.com/accounts-js/accounts/blob/master/packages/password/src/accounts-password.ts#L601, returning false should work properly.

**Reply by pradel** [reply]
can you share the configuration you are using for the project?

**Reply by MohanNarayana** [reply]
It is an open source eCommerce Node project. It has a distributed API architecture and the api plugin for authentication, where I am adding the code is [here](https://github.com/reactioncommerce/api-plugin-authentication/blob/trunk/src/util/accountServer.js#L28)
